### ISSUE TITLE
torch.utils._pytree -> stable

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#65761]
## 🚀 Feature

PyTorch should use a performant and stable version of _pytree.

## Motivation

`torch.utils._pytree` is something I cooked up when prototyping the initial non-composable version of vmap. Since then, it's found many legitimate use cases in PyTorch internals:
- functorch [transforms](https://github.com/facebookresearch/functorch/blob/6c026e965c50ab25a3a63f2625691c9a4a4a18ff/functorch/_src/eager_transforms.py#L13); [eager compilation](https://github.com/facebookresearch/functorch/blob/6c026e965c50ab25a3a63f2625691c9a4a4a18ff/functorch/_src/eager_compilation.py#L85)
- [fx](https://github.com/pytorch/pytorch/blob/master/torch/fx/_pytree.py)
- [writing torch_dispatch objects](https://github.com/pytorch/pytorch/blob/4752453d27c275df5f8efdea651ed9f08e366c7a/test/test_python_dispatch.py#L62)

The last use case means we want to expose torch.utils._pytree to users! (or provide a stable mechanism that users can use).

There is only really one main blocker to exposing torch.utils._pytree: performance.
- [tree_map is slow](https://github.com/pytorch/pytorch/issues/65753), it can probably be faster.
- Some of the eager compilation work is showing that torch.utils._pytree is slow, especially in the case where the input is a flat list.

## Pitch

torch.utils._pytree is heavily inspired by JAX's pytrees. JAX's pytrees are implemented in C++ (with bindings to Python) for performance reasons. We should either:
- delete our pytree and take a dependency on JAX (or figure out if they'd be interested in splitting off pytree into its own library)
- run our own implementation of pytree and improve its performance.

cc @albanD @Chillee @ezyang

**Comment by Chillee** [comment]
Some benchmarks

TL;DR: Deepmind's Pytree implementation is often an order of magnitude+ slower than Jax's for unflattening. This is likely due to the fact that Deepmind's pytree provides an `unflatten_as` API while Jax uses a flatten spec. Our current implementation is way slower in almost all situations :)
```
empty list f_pt:          0.001885890961 ms
empty list f_jax:         0.000846385956 ms
empty list f_dm:          0.001115798950 ms
empty list f_u_pt:        0.002555847168 ms
empty list f_u_jax:       0.002150535583 ms
empty list f_u_dm:        0.005831718445 ms

single list f_pt:         0.003507137299 ms
single list f_jax:        0.000922679901 ms
single list f_dm:         0.001225471497 ms
single list f_u_pt:       0.004913806915 ms
single list f_u_jax:      0.001626014709 ms
single list f_u_dm:       0.006451606750 ms

5 flat list f_pt:         0.011200904846 ms
5 flat list f_jax:        0.001156330109 ms
5 flat list f_dm:         0.000998973846 ms
5 flat list f_u_pt:       0.013060569763 ms
5 flat list f_u_jax:      0.002152919769 ms
5 flat list f_u_dm:       0.007913112640 ms

100 flat list f_pt:       0.159444808960 ms
100 flat list f_jax:      0.006237030029 ms
100 flat list f_dm:       0.008158683777 ms
100 flat list f_u_pt:     0.250461101532 ms
100 flat list f_u_jax:    0.009875297546 ms
100 flat list f_u_dm:     0.078125000000 ms

dict with 5 list entry f_pt:      0.025532245636 ms
dict with 5 list entry f_jax:     0.002255439758 ms
dict with 5 list entry f_dm:      0.001995563507 ms
dict with 5 list entry f_u_pt:    0.039303302765 ms
dict with 5 list entry f_u_jax:   0.003547668457 ms
dict with 5 list entry f_u_dm:    0.038719177246 ms

5 deep dict f_pt:         0.163798332214 ms
5 deep dict f_jax:        0.009055137634 ms
5 deep dict f_dm:         0.009317398071 ms
5 deep dict f_u_pt:       0.233585834503 ms
5 deep dict f_u_jax:      0.014815330505 ms
5 deep dict f_u_dm:       0.182533264160 ms
```

Benchmarked code
```
bench_inp("empty list", [])
bench_inp("single list", [torch.randn(5)])
bench_inp("5 flat list", [torch.randn(5) for _ in range(5)])
bench_inp("100 flat list", [torch.randn(5) for _ in range(100)])
bench_inp("dict with 5 list entry", {i: [torch.randn(5)] for i in range(5)})
def nest_deeper(x):
    return pytree.tree_map(lambda x: {'a': torch.randn(5), 'b': torch.randn(5)}, x)
nested_dict = [torch.randn(5)]
for _ in range(5):
    nested_dict = nest_deeper(nested_dict)
bench_inp("5 deep dict", nested_dict)
```

**Comment by cgarciae** [comment]
Hey! It would be amazing for the ecosystem if there was a single shared Pytree library, the concept is orthogonal to numeric computing and the whole Python community could benefit (e.g. I've installed JAX just to manipulate a nested dict).

I am creating this small [library](https://github.com/cgarciae/treeo) to create custom Pytrees, while currently it only works with JAX it would be nice if in the future it could be more universally useful. 

In the meanwhile, if there are going to be multiple Pytree libraries, it would be great if the was a common Pytree protocol that any class could implement to be picked up by all Pytree libraries without having to explicitly register to them. It could be something like:

```python
class Pytree(Protocol):

    def __tree_flatten__(self) -> Tuple[Sequence[Any], Any]:
        ...
    
    @classmethod
    def __tree_unflatten__(cls, aux: Any, children: Sequence[Any]) -> Any:
        ...
```

**Comment by ezyang** [comment]
One difficult thing is that fast pytree has to be done in C/C++ so the community library would be in the business of bundling a C extension and properly releasing binaries everywhere (e.g., on conda). Also, for maximum portability, it would probably be best if the extension was written in pure C, so that it interoperates smoothly with other binaries produced by the respective projects. But I do think the functionality of this library is narrowly scoped enough that maybe we can make it work.

**Comment by cgarciae** [comment]
I've started a discussion about a possible Pytree Protocol in google/jax#8099.

**Comment by Chillee** [comment]
I think the primary thing that's important to share is the C++ library. I think the python library is not that hard to implement (I feel like people implement versions of it all the time...)

**Comment by gchanan** [comment]
could this potentially be a header only thing in order to reduce the maintenance/release burden?  Or is the value-add of a python library shipped on Pypi too great here?

**Comment by IvanKobzarev** [comment]
I have a draft PR to support this: 
https://github.com/pytorch/pytorch/pull/75822

**Comment by zou3519** [comment]
Increasing priority due to user (developer) requests

**Comment by NicolasHug** [comment]
> Increasing priority due to user (developer) requests

For context: torchvision has started relying on `tree_map()`: https://github.com/pytorch/vision/pull/6419#issuecomment-1215091387

CC @pmeier @vfdev-5 @datumbox

**Comment by vfdev-5** [comment]
It also relies on  `_pytree.tree_flatten` and `_pytree.tree_unflatten`

**Comment by XuehaiPan** [comment]
Hi everyone, we are announcing our [`optree`](https://github.com/metaopt/optree), the Optimized PyTree Utilities than JAX's pytree. Please check out our [README](https://github.com/metaopt/optree) if you are interested.

### 0. Out-of-box compatible with JAX's pytree, provides the same interfaces and functions (and more).

### 1. High-performance: `optree` has comparable fast tree operations (~0.9x for `dict`s and ~2.5x for `OrderedDict`s) than JAX's pytree and it is 20x faster than `torch.utils._pytree`.

`optree` implements some common Python container types in C++ (e.g., `OrderedDict`) and achieves 2.5x performance than JAX's pytree. Check out section [Built-in PyTree Node Types](https://github.com/metaopt/optree#built-in-pytree-node-types) and [Benchmark](https://github.com/metaopt/optree#benchmark) for more details.

| Module    | Nodes | OpTree (μs) | JAX XLA (μs) | PyTorch (μs) | DM-Tree (μs) | Speedup (J / O) | Speedup (P / O) | Speedup (D / O) |
| :-------- | ----: | ----------: | -----------: | -----------: | -----------: | --------------: | --------------: | --------------: |
| TinyMLP   |    53 |       26.40 |        68.19 |       586.87 |        34.14 |            2.58 |           22.23 |            1.29 |
| AlexNet   |   188 |       84.28 |       259.51 |      2182.07 |       125.12 |            3.08 |           25.89 |            1.48 |
| ResNet18  |   698 |      288.57 |       807.27 |      7881.69 |       429.39 |            2.80 |           27.31 |            1.49 |
| ResNet34  |  1242 |      580.75 |      1564.97 |     15082.84 |       819.02 |            2.69 |           25.97 |            1.41 |
| ResNet50  |  1702 |      791.18 |      2081.17 |     20982.82 |      1104.62 |            2.63 |           26.52 |            1.40 |
| ResNet101 |  3317 |     1603.93 |      3939.37 |     40382.14 |      2208.63 |            2.46 |           25.18 |            1.38 |
| ResNet152 |  4932 |     2446.56 |      6267.98 |     56892.36 |      3139.17 |            2.56 |           23.25 |            1.28 |
| ViT-H/14  |  3420 |     1681.48 |      4488.33 |     41703.16 |      2504.86 |            2.67 |           24.80 |            1.49 |
| Swin-B    |  2881 |     1565.41 |      4091.10 |     34241.99 |      1936.75 |            2.61 |           21.87 |            1.24 |
|           |       |             |              |              |  **Average** |        **2.68** |       **24.78** |        **1.38** |

<div align="center">
  <img src="https://user-images.githubusercontent.com/16078332/200494435-fd5bb385-59f7-4811-b520-98bf5763ccf3.png" width="90%" />
</div>

### 2. Namespace Isolation for the PyTree Type Registry
 
In addition to the JAX's pytree registry for custom node type registration, `optree` add `namespace` isolation to the registry. Users can register the same type multiple times for different flatten/unflatten behavior. It also provides module-level isolation for safety reasons. For example, you can add a unique prefix to your namespace to isolate your registry with other modules (e.g., `torch.xxx`, `torch.functorch.xxx`):

```python
# Register a Python type into a namespace
import torch

optree.register_pytree_node(
    torch.Tensor,
    # (tensor) -> (children, metadata)
    flatten_func=lambda tensor: (
        (tensor.cpu().numpy(),),
        dict(dtype=tensor.dtype, device=tensor.device, requires_grad=tensor.requires_grad),
    ),
    # (metadata, children) -> tensor
    unflatten_func=lambda metadata, children: torch.tensor(children[0], **metadata),
    namespace='torch.torch2numpy',
)
```

```python
>>> tree = {'weight': torch.ones(size=(1, 2)).cuda(), 'bias': torch.zeros(size=(2,))}
>>> tree
{'weight': tensor([[1., 1.]], device='cuda:0'), 'bias': tensor([0., 0.])}

# Flatten without specifying the namespace
>>> tree_flatten(tree)  # `torch.Tensor`s are leaf nodes
([tensor([0., 0.]), tensor([[1., 1.]], device='cuda:0')], PyTreeSpec({'bias': *, 'weight': *}))

# Flatten with the namespace
>>> leaves, treespec = optree.tree_flatten(tree, namespace='torch.torch2numpy')
>>> leaves, treespec
(
    [array([0., 0.], dtype=float32), array([[1., 1.]], dtype=float32)],
    PyTreeSpec(
        {
            'bias': CustomTreeNode(Tensor[{'dtype': torch.float32, 'device': device(type='cpu'), 'requires_grad': False}], [*]),
            'weight': CustomTreeNode(Tensor[{'dtype': torch.float32, 'device': device(type='cuda', index=0), 'requires_grad': False}], [*])
        },
        namespace='torch.torch2numpy'
    )
)

# `entries` are not defined and use `range(len(children))`
>>> optree.tree_paths(tree, namespace='torch.torch2numpy')
[('bias', 0), ('weight', 0)]

# Unflatten back to a copy of the original object
>>> optree.tree_unflatten(treespec, leaves)
{'bias': tensor([0., 0.]), 'weight': tensor([[1., 1.]], device='cuda:0')}
```

Check out section [Registering a Container-like Custom Type as Non-leaf Nodes](https://github.com/metaopt/optree#notes-about-the-pytree-type-registry) for more details.

### 3. Support both `None` as Non-leaf Node and `None` as Leaf

In JAX's implementation, `None` is always an internal non-leaf node with an arity 0, which is like an empty tuple. This limits the usage of the JAX's pytree utilities for PyTorch. For example, the `nn.Module` uses `_parameters` and `_buffers` to hold the tensors, while the value can be a tensor or `None`.

`optree` supports both `None` as Non-leaf Node (JAX's default) and `None` as Leaf (PyTorch's default). Check out section [None is Non-leaf Node vs. None is Leaf](https://github.com/metaopt/optree#none-is-non-leaf-node-vs-none-is-leaf) for more details.

### 4. Some other improvements and bug fixes

1. Adds support for tree flatten and tree map with paths. (useful for `functorch` module extraction).
2. Improves the JAX's pytree sorting support for `dict`s.
3. Better string representation `repr(PyTreeSpec)`.
4. Fixes some bugs for JAX's pytree of hashing, pickle serialization, and tree-compose/tree-transpose.

**Comment by Chillee** [comment]
@XuehaiPan Exciting! We'll check it out and possibly replace our implementation with optree :)

**Comment by XuehaiPan** [comment]
Assign this to myself to make C++ pytree to be compatible with Dynamo.
